File: test/models/bookmark_test.js
const assert = require('assert');
const Dvd = require('../../models/dvd');
const Bookmark = require('../../models/bookmark_schema');

describe('Bookmarks on a Dvd', () => {
  it('can create a Bookmark', (done) => {
    const drStrange = new Dvd({title: 'Doctor Strange', bookmarks: [{name: 'Episode 1', time: 20}]});
    drStrange.save()
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        assert(dvd.bookmarks.length === 1);
        assert(dvd.bookmarks[0].time == 20);
        done();
      });
  });

  it('can add a Bookmark to an existing Dvd', (done) => {
    const drStrange = new Dvd({title: 'Doctor Strange', bookmarks: []});
    drStrange.save()
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        dvd.bookmarks.push({name: 'Episode 2', time: 90});
        return dvd.save()
      })
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        assert(dvd.bookmarks.length === 1);
        assert(dvd.bookmarks[0].time == 90);
        done();
      });
  });

  it('a Bookmark can be removed from a Dvd', (done) => {
    const drStrange = new Dvd({title: 'Doctor Strange', bookmarks: [{name: 'Episode 3', time: 130}]});
    drStrange.save()
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        dvd.bookmarks[0].remove()
        return dvd.save();
      })
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        assert(dvd.bookmarks.length == 0);
        done();
      });
  });

  it('can add a Bookmark to an Episode', (done) => {
    const drStrange = new Dvd({title: 'Doctor Strange', episodes: [{name: 'Disc 1'}], bookmarks: []});
    drStrange.save()
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        dvd.episodes[0].bookmarks.push({name: 'Episode 9', time: 450});
        return dvd.save()
      })
      .then(() => Dvd.findOne({title: 'Doctor Strange'}))
      .then((dvd) => {
        assert(dvd.episodes[0].bookmarks.length === 1);
        assert(dvd.episodes[0].bookmarks[0].time == 450);
        done();
      });
  });
});

File: models/bookmark_test.go
package models

import (
"testing"
"dvdpila/db"
"flag"
)

var episode Episode

func init() {
	ldb = db.Connect()
	if flag.Lookup("test.v") != nil {
		CreateTables(ldb)
	}

	dvd = Dvd{
		Title:        "Ghost In The Shell",
		Rating:       5,
		AbstractTxt:  `In the near future, the vast majority of humans are [[Human enhancement|augmented]] with cybernetics, enhancing various traits like vision, strength, and intelligence. Hanka Robotics, the world's leading developer of augmentative technology, establishes a secret project to develop a mechanical body, or "shell", that can [[Brain–computer interface|integrate]] a human brain rather than an [[artificial intelligence|AI]]. `,
		AbstractSrc:  "Wikipedia",
		AbstractURL:  "https://en.wikipedia.org/wiki/Ghost_in_the_Shell_(2017_film)",
		ImageURL:     "http://dvdpila/images/posters/Ghost_in_the_Shell_2017_film.png",
		FileURL:      "http://videos/Ghost_In_The_Shell_2017.mkv",
		PlaybackTime: 0,
	}
	ldb.Create(&dvd)

	episode = Episode{
		Name: "Disc 1",
		EpisodeFileURL: "http://videos/Ghost_In_The_Shell_2017.mkv",
		PlaybackTime: 0,
		Dvd: dvd,
		DvdID: dvd.ID,
	}
	ldb.Create(&episode)
}


func TestCreateDeleteBookmark(t *testing.T) {
	bookmark := Bookmark{
		Name: "Funny Thing",
		Time: 20,
		Dvd: dvd,
		DvdID: dvd.ID,
	}
	ldb.Create(&bookmark)

	var bk Bookmark
	ldb.First(&bk, bookmark.ID)

	if (bk.Name != "Funny Thing") {
		t.Errorf("Expected bk.Name to be 'Funny Thing' but it is %v", bk.Name)
	}

	ldb.Delete(&bk)

	bkId := bookmark.ID
	var bks Bookmark
	ldb.First(&bks, bkId)

	if (bks.Name != "") {
		t.Errorf("Expected bks.Name to be '' but it is not...")
	}
}

func TestBookmarkBelongsToDvd(t *testing.T) {
	bookmark := Bookmark{
		Name: "Funny Thing",
		Time: 25,
		Dvd: dvd,
		DvdID: dvd.ID,
	}
	ldb.Create(&bookmark)

	var bk Bookmark
	ldb.First(&bk, bookmark.ID)
	ldb.Model(&bk).Related(&bk.Dvd)

	if (bk.Dvd.Title != "Ghost In The Shell") {
		t.Errorf("Expected ep.Dvd.Title to be 'Ghost In The Shell' but it is %v", bk.Dvd.Title)
	}
}

func TestBookmarkBelongsToEpisode(t *testing.T) {
	bookmark := Bookmark{
		Name: "Funny Thing",
		Time: 25,
		Episode: episode,
		EpisodeID: episode.ID,
	}
	ldb.Create(&bookmark)

	var bk Bookmark
	ldb.First(&bk, bookmark.ID)
	ldb.Model(&bk).Related(&bk.Episode)

	if (bk.Episode.Name != "Disc 1") {
		t.Errorf("Expected ep.Episode.Name to be 'Disc 1' but it is %v", bk.Episode.Name)
	}
}